File: tbrpgsca/src/com/tbrpgsca/library/Actor.java
/*
Copyright (C) 2015 Claudiu-Stefan Costea

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. 
*/
package com.tbrpgsca.library;

import java.util.ArrayList;

public class Actor extends Job {
	public String name;
	public boolean active = true, reflect = false;
	public State state[];
	public int auto = 0, hp, mp, sp, level, maxlv, exp, maxp, atk, def, spi, wis, agi;
	public int mres[], res[] = { 3, 3, 3, 3, 3, 3, 3, 3 };

	public ArrayList<Integer> jobSkills = new ArrayList<Integer>();

	public void setRace(Race race, Boolean init) {
		if (init) {
			this.raceStats(race.maxhp, race.maxmp, race.maxsp, race.matk, race.mdef, race.mwis, race.mspi, race.magi);
			this.stats(1, this.maxlv);
		} else
			this.raceStats(this.maxhp + race.maxhp, this.maxmp + race.maxmp, this.maxsp + race.maxsp,
					this.matk + race.matk, this.mdef + race.mdef, this.mwis + race.mwis, this.mspi + race.mspi,
					this.magi + race.magi);
		this.setRes(race.rres, true);
		this.raceSkills = race.raceSkills;
		this.rname = race.rname;
	}

	public void changeJob(Job job, Boolean add) {
		this.jname = job.jname;
		if (add)
			this.raceStats(this.maxhp + job.maxhp, this.maxmp + job.maxmp, this.maxsp + job.maxsp, this.matk + job.matk,
					this.mdef + job.mdef, this.mwis + job.mwis, this.mspi + job.mspi, this.magi + job.magi);
		this.jobStats(job.hpp, job.mpp, job.spp, job.atkp, job.defp, job.wisp, job.spip, job.agip);
		if (!add) {
			this.removeSkills();
			this.resetRes();
		}
		this.setRes(job.rres, false);
		this.addSkills(job.raceSkills);
		this.setSprName(job.getSprName());
	}

	public void checkRes(int r) {
		if (r < 8 && r >= 0) {
			if (this.rres[r] > 7)
				this.rres[r] = 7;
			if (this.rres[r] < 0)
				this.rres[r] = 0;
			if (this.mres[r] > 7)
				this.mres[r] = 7;
			if (this.mres[r] < 0)
				this.mres[r] = 0;
			if (this.res[r] > 7)
				this.res[r] = 7;
			if (this.res[r] < 0)
				this.res[r] = 0;
		}
	}

	public String applyState(boolean consume) {
		String s = "";
		if (auto < 2 && auto > -2)
			this.auto = 0;
		else
			this.auto = 2;
		if (consume && hp > 0)
			active = true;
		this.atk = this.matk;
		this.def = this.mdef;
		this.spi = this.mspi;
		this.wis = this.mwis;
		this.agi = this.magi;
		for (int i = 0; i < this.res.length; i++)
			this.res[i] = this.mres[i];
		this.reflect = false;
		boolean c = false;
		for (int i = 0; i < this.state.length; i++)
			if (this.state[i].dur != 0 && this.hp > 0) {
				String r = state[i].apply(this, consume);
				if (r.length() > 0) {
					if (c)
						s += ", ";
					if (consume && !c) {
						s += "\n";
						c = true;
					}
					s += r;
				}
			}
		s += checkStatus();
		if (c && consume)
			s += ".";
		return s;
	}

	public String checkStatus() {
		String s = "";
		if (this.hp > maxhp)
			hp = maxhp;
		if (this.mp > maxmp)
			mp = maxmp;
		if (this.sp > maxsp)
			sp = maxsp;
		if (this.hp < 1) {
			s += " (and falls unconcious)";
			active = false;
			for (int i = 0; i < state.length; i++)
				this.state[i].remove();
		}
		if (this.hp < -maxhp)
			hp = -maxhp;
		if (this.mp < 0)
			mp = 0;
		if (this.sp < 0)
			sp = 0;
		return s;
	}

	public Actor() {
		this("Actor", 5);
	}

	public Actor(String name, int maxlv) {
		this.name = name;
		this.state = this.AddStates();
		this.stats(1, maxlv);
	}

	public Actor(String name, String race, String jname, int lv, int maxlv, int maxhp, int maxmp, int maxsp, int atk,
			int def, int wis, int spi, int agi, int hpp, int mpp, int spp, int atkp, int defp, int wisp, int spip,
			int agip) {
		this(name, race, jname, lv, maxlv, maxhp, maxmp, maxsp, atk, def, wis, spi, agi, hpp, mpp, spp, atkp, defp,
				wisp, spip, agip, new int[] {});
	}

	public Actor(String name, String race, String jname, int lv, int maxlv, int maxhp, int maxmp, int maxsp, int atk,
			int def, int wis, int spi, int agi, int hpp, int mpp, int spp, int atkp, int defp, int wisp, int spip,
			int agip, int[] newRes) {
		this(name, race, jname, lv, maxlv, maxhp, maxmp, maxsp, atk, def, wis, spi, agi, hpp, mpp, spp, atkp, defp,
				wisp, spip, agip, newRes, new int[] {});
	}

	public Actor(String name, String race, String jname, int lv, int maxlv, int maxhp, int maxmp, int maxsp, int atk,
			int def, int wis, int spi, int agi, int hpp, int mpp, int spp, int atkp, int defp, int wisp, int spip,
			int agip, int[] newRes, int[] newSkill) {
		super(jname, maxhp, maxmp, maxsp, atk, def, wis, spi, agi, hpp, mpp, spp, atkp, defp, wisp, spip, agip, newRes,
				newSkill);
		this.name = name;
		this.rname = race;
		this.state = this.AddStates();
		this.raceSkills = newSkill;
		this.stats(lv, maxlv);
	}

	private void setRes(int nres[], boolean race) {
		for (int i = 0; i < this.res.length; i++) {
			if (race)
				this.rres[i] = nres[i];
			else
				this.mres[i] += nres[i];
			this.checkRes(i);
		}
		if (race) {
			this.resetRes();
			this.setRes(nres, false);
		}
	}

	private void resetRes() {
		for (int i = 0; i < this.mres.length; i++)
			this.mres[i] = 3 + rres[i];
	}

	private void stats(int lv, int maxlv) {
		this.resetRes();
		this.maxlv = maxlv;
		this.maxp = 5;
		if (lv > 1) {
			this.exp = (int) (this.maxp * (Math.pow(lv - 1, 2)));
		} else
			this.exp = 0;
		this.level = 1;
		this.levelUp();
		this.recover();
		if (this.hp < 1)
			this.active = false;
	}

	public void recover() {
		this.hp = this.maxhp;
		this.mp = this.maxmp;
		this.sp = this.maxsp;
		this.active = true;
		this.atk = this.matk;
		this.def = this.mdef;
		this.spi = this.mspi;
		this.wis = this.mwis;
		this.agi = this.magi;
		for (int i = 0; i < this.state.length; i++)
			this.state[i].remove();
		this.applyState(false);
	}

	public void copy(Actor cloned) {
		this.name = cloned.name;
		this.jname = cloned.jname;
		this.rname = cloned.rname;
		this.raceStats(cloned.maxhp, cloned.maxmp, cloned.maxsp, cloned.atk, cloned.def, cloned.wis, cloned.spi,
				cloned.agi);
		this.jobStats(cloned.hpp, cloned.mpp, cloned.spp, cloned.atkp, cloned.defp, cloned.wisp, cloned.spip,
				cloned.agip);
		this.stats(cloned.level, cloned.maxlv);
		for (int i = 0; i < state.length; i++) {
			this.state[i].dur = cloned.state[i].dur;
			this.state[i].res = cloned.state[i].res;
		}
		for (int i = 0; i < res.length; i++) {
			this.rres[i] = cloned.rres[i];
			this.mres[i] = cloned.mres[i];
			this.res[i] = cloned.res[i];
		}
		this.jobSkills.clear();
		this.raceSkills = cloned.raceSkills;
		this.addSkills(cloned.jobSkills);
		this.setSprName(cloned.getSprName());
	}

	public void addSkills(ArrayList<Integer> newSkill) {
		for (int i = 0; i < newSkill.size(); i++)
			if (!this.jobSkills.contains(newSkill.get(i)))
				this.jobSkills.add(newSkill.get(i));
	}

	@Override
	public void setResistance(int[] newRes) {
		this.mres = new int[] { 3, 3, 3, 3, 3, 3, 3, 3 };
		for (int i = 0; i < newRes.length && i < this.mres.length; i++)
			this.mres[i] = newRes[i];
	}

	private State[] AddStates() {
		State state[] = new State[11];

		state[0] = new State("Regen", false, false, false, -1, 10, 0, 0, 0, 2, 0, 0, 0, false);
		state[1] = new State("Poison", false, false, false, 10, -7, 0, -2, 0, -2, 0, 0, 0, false);
		state[2] = new State("Clarity", false, false, false, -1, 0, 7, 0, 0, 0, 1, 1, 0, false);
		state[3] = new State("Dizziness", false, false, false, 3, 0, -7, 0, 0, 0, -1, -1, 0, false);
		state[4] = new State("Vigour", false, false, false, -1, 0, 0, 7, 1, 0, 0, 0, 1, false);
		state[5] = new State("Weakness", false, false, false, 5, 0, 0, -7, -1, 0, 0, 0, -1, false);
		state[6] = new State("Berserk", false, true, false, 7, 0, 0, 0, 5, -3, 0, 0, 3, false);
		state[7] = new State("Confusion", false, false, true, 3, 0, 0, 0, 0, 0, 0, 0, 0, false);
		state[8] = new State("Sleep", true, false, false, 5, 0, 0, 0, 0, -3, 0, 0, -3, false);
		state[9] = new State("Stun", true, false, false, 1, 0, 0, 0, 0, -1, 0, 0, -1, false);
		state[10] = new State("Reflect", false, false, false, 7, 0, 0, 0, 0, 0, 0, 0, 0, true);

		return state;
	}

	public void levelUp() {
		while (this.maxp <= this.exp && this.level < this.maxlv) {
			this.maxp *= 2;
			this.level++;
			this.maxhp += 3;
			this.maxmp += 2;
			this.maxsp += 2;
			this.matk++;
			this.mdef++;
			this.mwis++;
			this.mspi++;
			this.magi++;
			if (this.level % 2 == 1) {
				this.maxhp += this.hpp;
				this.maxmp += this.mpp;
				this.maxsp += this.spp;
				this.matk += this.atkp;
				this.mdef += this.defp;
				this.mwis += this.wisp;
				this.mspi += this.spip;
				this.magi += this.agip;
			}
		}
	}

	public class State {
		public boolean inactive, confusion, auto, reflect;
		public String name;
		public int res, mdur, dur, hpm, mpm, spm, atkm, defm, spim, wism, agim;
		public int resm[] = { 0, 0, 0, 0, 0, 0, 0, 0 };

		public State(String name, boolean inactive, boolean auto, boolean confusion, int dur, int hpm, int mpm, int spm,
				int atkm, int defm, int wism, int spim, int agim, boolean reflect) {
			this.name = name;
			this.mdur = dur;
			this.dur = 0;
			this.res = 0;
			this.hpm = hpm;
			this.mpm = mpm;
			this.spm = spm;
			this.atkm = atkm;
			this.defm = defm;
			this.spim = spim;
			this.wism = wism;
			this.agim = agim;
			this.inactive = inactive;
			this.confusion = confusion;
			this.auto = (confusion) ? true : auto;
			this.reflect = reflect;
		}

		public void inflict() {
			int rnd = (int) (Math.random() * 10);
			if (this.dur > -2 && rnd > res)
				this.dur = this.mdur;
		}

		public String apply(Actor actor, boolean consume) {
			String s = "";
			if (this.dur != 0 && actor.hp > 0) {
				if (consume) {
					int rnd = (int) (Math.random() * 3);
					int dmghp = ((actor.maxhp + rnd) * this.hpm) / 100;
					int dmgmp = ((actor.maxmp + rnd) * this.mpm) / 100;
					int dmgsp = ((actor.maxsp + rnd) * this.spm) / 100;
					actor.hp += dmghp;
					actor.mp += dmgmp;
					actor.sp += dmgsp;
					boolean c = false;
					if (dmghp != 0 || dmgmp != 0 || dmgsp != 0)
						s += (name + " causes " + actor.name);
					if (dmghp != 0) {
						s += (" ");
						if (dmghp >= 0)
							s += "+";
						s += dmghp + " HP";
						c = true;
					}
					if (dmgmp != 0) {
						if (c)
							s += ",";
						s += " ";
						if (dmgmp >= 0)
							s += "+";
						s += dmgmp + " MP";
						c = true;
					}
					if (dmgsp != 0) {
						if (c)
							s += ",";
						s += " ";
						if (dmgsp >= 0)
							s += "+";
						s += dmgsp + " SP";
					}
					if (this.dur > 0)
						this.dur--;
				}
				actor.atk = actor.matk + this.atkm;
				actor.def = actor.mdef + this.defm;
				actor.spi = actor.mspi + this.spim;
				actor.wis = actor.mwis + this.wism;
				actor.agi = actor.magi + this.agim;
				for (int i = 0; i < actor.res.length; i++) {
					actor.res[i] += this.resm[i];
					actor.checkRes(i);
				}
				if (this.inactive)
					actor.active = false;
				if (this.reflect)
					actor.reflect = true;
				if (this.auto && actor.auto < 2)
					actor.auto = 1;
				if (this.confusion)
					actor.auto = (actor.auto < 2) ? -1 : -2;
			}
			return s;
		}

		public void remove() {
			if (this.dur > -2)
				this.dur = 0;
		}
	}

	public void addSkills(int[] newSkill) {
		for (int j = 0; j < newSkill.length; j++)
			if (!this.jobSkills.contains(newSkill[j]))
				this.jobSkills.add(newSkill[j]);
	}

	public void removeSkills() {
		this.jobSkills.clear();
	}

}

File: ktbrpgsca/app/src/main/java/com/codycostea/tbrpgsca/Actor.kt
/*
Copyright (C) AD 2018 Claudiu-Stefan Costea

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package com.codycostea.tbrpgsca

import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashMap

open class Actor(id : Int, name: String, race: Costume, job: Costume, level : Int = 1, open var maxLv: Int = 9,
                 mActions : Int = 1, mHp: Int = 30, mMp: Int = 10, mSp: Int = 10, mAtk: Int = 7, mDef: Int = 7,
                 mSpi: Int = 7, mWis: Int = 7, mAgi: Int = 7, range : Boolean = false, mRes: MutableMap<Int, Int>? = null,
                 skills: Array<Ability>? = null, states: Array<State>? = null, mStRes: MutableMap<State, Int>? = null)
    : Costume(id, name, mHp + race.mHp + job.mHp, mMp + race.mMp + job.mMp, mSp + race.mSp + job.mSp,
        mAtk + race.atk + job.atk, mDef + race.def + job.def, mSpi + race.spi + job.spi,
        mWis + race.wis + job.wis, mAgi + race.agi + job.agi, mActions, range, mRes, skills, states, mStRes) {

    companion object {
        var koTxt = ", %s falls unconscious"
    }

    open var race : Costume = race
        set(value) {
            this.switchCostume(field, value)
            field = value
        }

    open var job : Costume = job
        set(value) {
            this.switchCostume(field, value)
            field = value
        }

    private var _lv : Int = 1
    open var level : Int
        get() = this._lv
        set(value) {
            var v = value
            if (value > this.maxLv) {
                v = this.maxLv
            }
            while (this._lv < v) {
                this.exp = this.mExp
                this.levelUp()
            }
        }

    open var exp : Int = (level - 1) * 15
        set(value) {
            field = value
            this.levelUp()
        }
    open var mExp : Int = level * 15

    open var actions : Int = this.mActions
    open var guards : Boolean = true
    open var reflect : Boolean = false
    open var automatic : Int = 0

    open var hp : Int = this.mHp
        set(value) {
            field = when {
                value > this.mHp -> this.mHp
                value < 0 -> 0
                else -> value
            }
        }
    open var mp : Int = this.mMp
        set(value) {
            field = when {
                value > this.mMp -> this.mMp
                value < 0 -> 0
                else -> value
            }
        }
    open var sp : Int = this.mSp
        set(value) {
            field = when {
                value > this.mSp -> this.mSp
                value < 0 -> 0
                else -> value
            }
        }

    open var stateDur : MutableMap<State, Int>? = null
        internal set

    open val availableSkills : ArrayList<Ability> = ArrayList()

    private var ranged: Boolean? = null
    override var range: Boolean
        get() {
            var ranged = this.ranged
            if (ranged === null) {
                val equipment = this.equipment?.values
                val states = this.stateDur
                ranged = super.range || this.job.range || this.race.range
                        || (equipment !== null && equipment.any { it.range }
                        || (states !== null && states.any { it.value != 0 && it.key.range }))
                this.ranged = ranged
            }
            return ranged
        }
        set(value) {
            super.range = value
            if (value) {
                this.ranged = true
            }
            else {
                this.ranged = null
            }
        }

    internal var skillsQty : MutableMap<Ability, Int>? = null
    internal var skillsQtyRgTurn : MutableMap<Ability, Int>? = null

    open internal var _items : LinkedHashMap<Ability, Int>? = null
    val items : LinkedHashMap<Ability, Int>
        get() {
            var items = this._items
            if (items === null) {
                items = LinkedHashMap()
                this._items = items
            }
            return items
        }

    private var equipment : MutableMap<Char, Costume>? = null

    open val equippedItems : Map<Char, Costume>?
        get() {
            return if (this.equipment === null) {
                null
            }
            else {
                HashMap<Char, Costume>(this.equipment)
            }
        }

    open fun equipItem(pos: Char, item : Costume) : Costume? {
        val r : Costume? = this.unequipPos(pos)
        var e = this.equipment
        if (e === null) {
            e = HashMap()
            this.equipment = e
        }
        e[pos] = item
        this.switchCostume(null, item)
        return r
    }

    open fun unequipPos(pos : Char) : Costume? {
        val e = this.equipment
        return if (e === null) {
            null
        }
        else {
            val r: Costume? = e.get(pos)
            this.switchCostume(r, null)
            r
        }
    }

    open fun unequipItem(item : Costume) : Char? {
        val e = this.equipment
        if (e !== null) {
            for (k in e.keys) {
                if (this.unequipPos(k) == item) {
                    return k
                }
            }
        }
        return null
    }

    private fun checkRegSkill(skill : Ability) {
        if (skill.rQty > 0) {
            var regSkills = this.skillsQtyRgTurn
            if (regSkills === null) {
                regSkills = HashMap()
                this.skillsQtyRgTurn = regSkills
            }
            regSkills[skill] = 0
        }
    }

    internal fun updateStates(remove : Boolean, states : Array<State>?) {
        if (states === null) {
            return
        }
        val aStates = this.stateDur//this.activeStates
        if (remove) {
            if (aStates === null) {
                return
            }
            for (k in states) {
                k.remove(this, true, true)
            }
        }
        else {
            for (k in states) {
                k.inflict(this, true, true)
            }
        }
    }

    internal fun updateSkills(remove : Boolean, abilities : Array<Ability>?) {
        if (abilities === null) {
            return
        }
        if (remove) {
            for (k in abilities) {
                this.availableSkills.remove(k)
                if (k.rQty > 0) {
                    val regSkills = this.skillsQtyRgTurn
                    if (regSkills !== null) {
                        regSkills.remove(k)
                    }
                }
            }
        }
        else {
            this.availableSkills.ensureCapacity(this.availableSkills.size + abilities.size)
            for (k in abilities) {
                this.availableSkills.add(k)
                this.checkRegSkill(k)
            }
        }
    }

    protected open fun switchCostume(oldRole : Costume?, newRole : Costume?) {
        if (oldRole !== null) {
            this.updateSkills(true, oldRole.skills)
            this.updateAttributes(true, oldRole)
            this.updateResistance(true, oldRole.res, oldRole.stRes)
            this.updateStates(true, oldRole.states)
        }
        if (newRole !== null) {
            this.updateStates(false, newRole.states)
            this.updateResistance(false, newRole.res, newRole.stRes)
            this.updateAttributes(false, newRole)
            this.updateSkills(false, newRole.skills)
        }
    }

    internal open fun updateAttributes(remove : Boolean, role : Costume) {
        val i: Int
        if (remove) {
            i = -1
            if (role.range) {
                this.ranged = null
            }
        }
        else {
            if (role.range) {
                this.ranged = true
            }
            i = 1
        }
        this.mHp += i * role.mHp
        this.mMp += i * role.mMp
        this.mSp += i * role.mSp
        this.atk += i * role.atk
        this.def += i * role.def
        this.mSp += i * role.spi
        this.wis += i * role.wis
        this.agi += i * role.agi
        this.mActions += i * role.mActions
    }

    internal fun updateResistance(remove : Boolean, resMap : Map<Int, Int>?, stResMap : Map<State, Int>?) {
        if (resMap !== null) {
            var r = this.res
            if (r === null) {
                if (remove) {
                    return
                }
                else {
                    r = HashMap(resMap.size)
                    this.res = r
                }
            }
            for (i in resMap.keys) {
                for (k in r.keys) {
                    if (!remove || i == k) {
                        r[k] = (r[k] ?: 3) + ((if (remove) -1 else 1) * (resMap[i] ?: 0))
                    }
                }
            }
        }
        if (stResMap !== null) {
            var rs = this.stRes
            if (rs === null) {
                if (remove) {
                    return
                }
                else {
                    rs = HashMap()
                    this.stRes = rs
                }
            }
            for (i in stResMap.keys) {
                for (k in rs.keys) {
                    if (!remove || i == k) {
                        rs[k] = (rs[k] ?: 3) + ((if (remove) -1 else 1) * (stResMap[i] ?: 0))
                    }
                }
            }
        }
    }

    fun levelUp() {
        while (this.mExp <= this.exp && this._lv < this.maxLv) {
            this.mExp *= 2
            this._lv++
            this.mHp += 3
            this.mMp += 2
            this.mSp += 2
            this.atk++
            this.def++
            this.wis++
            this.spi++
            this.agi++
        }
    }

    fun applyStates(consume: Boolean): String {
        var s = ""
        if (!consume) {
            if (this.automatic < 2 && this.automatic > -2) {
                this.automatic = 0
            }
            else {
                this.automatic = 2
            }
            if (this.hp > 0) {
                //if (consume) this.actions = this.mActions
                this.guards = true
            }
            this.reflect = false
        }
        var c = false
        val sDur = this.stateDur
        if (sDur !== null) {
            for (state in sDur) {
                if (state.value > -3 && this.hp > 0) {
                    val r = state.key.apply(this, consume)
                    if (r.isNotEmpty()) {
                        if (c) s += ", "
                        if (consume && !c) {
                            //s += "\n"
                            c = true
                        }
                        s += r
                    }
                }
            }
        }
        s += checkStatus()
        if (c && consume) s += "."
        return s
    }

    fun checkStatus(): String {
        var s = ""
        if (this.hp < 1) {
            s += String.format(Actor.koTxt, this.name)
            this.actions = 0
            this.guards = false
            this.sp = 0
            val sDur = this.stateDur
            if (sDur !== null) {
                for (state in sDur.keys) {
                    state.remove(this, false, false)
                }
            }
        }
        return s
    }

    fun recover() {
        this.hp = this.mHp
        this.mp = this.mMp
        this.sp = 0
        this.actions = this.mActions
        val sDur = this.stateDur
        if (sDur !== null) {
            for (state in sDur.keys) {
                state.remove(this, true, false)
            }
            if (sDur.isEmpty()) {
                this.stateDur = null
            }
        }
        this.applyStates(false)
        val res = this.res
        if (res !== null) {
            for (r in res) {
                if (r.value == 3) {
                    res.remove(r.key)
                }
            }
            if (res.isEmpty()) {
                this.res = null
            }
        }
        val stRes = this.stRes
        if (stRes !== null) {
            for (r in stRes) {
                if (r.value == 0) {
                    stRes.remove(r.key)
                }
            }
            if (stRes.isEmpty()) {
                this.stRes = null
            }
        }
        val skillQty = this.skillsQty
        if (skillQty !== null) {
            for (ability in skillQty.keys) {
                skillQty[ability] = ability.mQty
            }
        }
    }

    init {
        val raceSkills = race.skills
        val jobSkills = job.skills
        val s : Int = (raceSkills?.size ?: 0)
        + (jobSkills?.size ?: 0)
        + (skills?.size ?: 0)
        this.availableSkills.ensureCapacity(s)
        if (raceSkills !== null) {
            for (a in raceSkills) {
                this.checkRegSkill(a)
                this.availableSkills.add(a)
            }
        }
        if (jobSkills !== null) {
            for (a in jobSkills) {
                this.checkRegSkill(a)
                this.availableSkills.add(a)
            }
        }
        if (skills !== null) {
            for (a in skills) {
                this.checkRegSkill(a)
                this.availableSkills.add(a)
            }
        }
        this.updateStates(false, race.states)
        this.updateStates(false, job.states)
        this.updateStates(false, states)
        /*val e = this.equipment
        if (e != null) {
            for (k in e.keys) {
                this.updateStates(false, e[k]?.states)
            }
        }*/
        if (level > 1) {
            this.level = level
        }
    }
}